harness/determined/_preemption.py

```diff
@@ -0,0 +1,142 @@
+               print('finished without preemption signal')
+    """
+
+    def __init__(self, master_url: str, trial_id: int, cert: certs.Cert) -> None:
+        self._master_url = master_url
+        self._trial_id = trial_id
+        self._cert = cert
+
+        # We will use a selector loop as a non-blocking way to stream through the requests library.
+        # We use a socketpair for a cross-platform way to wake up a selector loop mid-stream.
+        self._rsock = None  # type: Optional[socket.socket]
+        self._wsock = None  # type: Optional[socket.socket]
+
+        self._should_preempt = None  # type: Optional[bool]
+
+        self._cond = threading.Condition()
+        super().__init__()
+
+    def run(self) -> None:
+        assert self._rsock is not None
+        assert self._wsock is not None
+
+        with api.get(
+            self._master_url,
+            f"/api/v1/trials/{self._trial_id}/signals/preemption",
+            stream=True,
+            cert=self._cert,
+        ) as r:
+            lines = r.iter_lines()
+
+            # The first response is always immediate, so we always block for it.  This is doubly
+            # important because we can only select on the HTTP connection socket _after_ we call
+            # next(lines) at least once.
+            for line in lines:
+                if not line:
+                    # Empty keepalive message.
+                    continue
+                break
+
+            j = json.loads(line)
+
+            with self._cond:
+                self._should_preempt = j["result"]["preempt"]
+                # Wake the main thread in case it was waiting for the main response.
+                self._cond.notify()
+
+            if self._should_preempt:
+                return
+
+            # It may be a long time between the first response and the second.  We want to be
+            # responsive to .close() calls from the main thread, so we block in input from either
+            # the HTTP connection socket or on our own _rsock socket, over which we can be alerted
+            # to .close() calls.
+            with selectors.DefaultSelector() as sel:
+                sel.register(r.raw, selectors.EVENT_READ)
+                sel.register(self._rsock, selectors.EVENT_READ)
+
+                while True:
+                    for key, _ in sel.select():
+                        if key.fileobj == self._rsock:
+                            # self._rsock is only written to when we are supposed to close.
+                            return
+
+                        assert key.fileobj == r.raw
+                        # Calling next() is blocking, but it should only block for a trivial amount
+                        # of time; basically only in the case that one message from the preemption
+                        # API got broken into two network packets.
+                        # r.raw._connection.sock.setblocking(True)
+                        line = next(lines)
```

Comment:
there's a temptation to just read from `r.raw` and not do `iter_lines()` at all. then you'd have at least some degree of control over `timeout`s on these read operations.

Reply:
r.raw is a `HTTPConnection` object, but it does have a `fileno()` method, which is why you can pass it to select